### ISSUE TITLE
wrong derivatives?

### ISSUE THREAD

**Issue by koen-dejonghe** [noelwelsh/functional-differential-programming#1]
Taking below snippet as reference:
```
  val x = Reverse(-2.0)
  val y = Reverse(5.0)
  val z = Reverse(-4.0)

  val q = x + y
  val f = q * z

  println(x.derivative)
  println(y.derivative)
  println(z.derivative)
  println(q.derivative)
  println(f.derivative)
```
I would expect the gradients to be as follows:
```
x.derivative == -4
y.derivative == -4
z.derivative == 3
q.derivative == -4
f.derivative == 1
```
See http://cs231n.github.io/optimization-2/#backprop

However, I get:
```
1.0
1.0
1.0
2.0
-5.0
```

**Comment by noelwelsh** [comment]
`fdl` is univariate. It doesn't support multivariate calculus.

**Comment by koen-dejonghe** [comment]
ok, I was not aware of that.
You may want to have a look at scorch where I have implemented autograd in a similar way as Pytorch. Maybe there is a more Scala-esque way if doing this? Much appreciated.

https://github.com/botkop/scorch/tree/master/src/main/scala/scorch/autograd

**Comment by noelwelsh** [comment]
The implementation technique you've used in Scorch (reification of the
expression language) is basically the only one that makes sense for a
complete system, unless you have a compiler plugin / macro that allows
compilation of source code into some other representation. LMS (
https://scala-lms.github.io/) is a project that enables the latter but I
would hesitate to use it in production code as it lags behind the main
Scala compiler, and I'm not clear how supported it is.
